torch/distributed/_tensor/ops/embedding_ops.py

```diff
@@ -1,97 +1,168 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
 # implement matrix related ops for distributed tensor
+import itertools
+from typing import cast, List
 
 import torch
-from torch.distributed._tensor.op_schema import OpSchema, OutputSharding
-from torch.distributed._tensor.ops.utils import register_prop_rule
+from torch.distributed._tensor.op_schema import (
+    OpSchema,
+    OpStrategy,
+    PlacementStrategy,
+    StrategyType,
+)
+from torch.distributed._tensor.ops.utils import (
+    generate_redistribute_costs,
+    is_tensor_shardable,
+    register_op_strategy,
+)
 
 from torch.distributed._tensor.placement_types import (
     _Partial,
     DTensorSpec,
+    Placement,
     Replicate,
     Shard,
 )
 
+from torch.distributed.device_mesh import DeviceMesh
+
 aten = torch.ops.aten
 
 
-# TODO: Enable BWD for embedding op.
-@register_prop_rule(aten.embedding.default)
-def embedding_rules(op_schema: OpSchema) -> OutputSharding:
-    weight_spec, inp_spec = op_schema.args_spec
+@register_op_strategy(aten.embedding.default)
+def embedding_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    """
+    This strategy handles embedding op. We have two possible embedding shardings:
+    rowwise and colwise
+    # TODO: implement rowwise sharding
+    """
+    weight_strategy = cast(OpStrategy, op_schema.args_schema[0])
+    indices_strategy = cast(OpStrategy, op_schema.args_schema[1])
+
+    weight_shape = weight_strategy.output_shape
+    indices_shape = indices_strategy.output_shape
+    output_emd_dim = len(indices_shape)
+
+    # guard rowwise sharding not implemented for now
+    weight_spec = weight_strategy.strategies[0].out_spec
     if any(placement.is_shard(0) for placement in weight_spec.placements):
         raise NotImplementedError(
             "DTensor does not support row-wise sharded embedding operation yet!"
         )
 
-    if weight_spec.is_replicated() and inp_spec.placements == [Shard(0)]:
-        # Embedding table is replicated, input ids are sharded along batch
-        # dimension. Output lookups should match input sharding spec in this case.
-        return OutputSharding(
-            output_spec=DTensorSpec(mesh=inp_spec.mesh, placements=inp_spec.placements)
-        )
+    all_mesh_dim_strategies = []
 
-    if inp_spec.is_replicated():
-        weight_dim_map = weight_spec.dim_map
-        output_dim_map = inp_spec.dim_map
-        output_dim_map.append(weight_dim_map[1])
-        return OutputSharding(
-            output_spec=DTensorSpec.from_dim_map(inp_spec.mesh, output_dim_map, [])
-        )
+    for mesh_dim in range(mesh.ndim):
+        single_mesh_dim_strategies = []
+
+        # placement list stores placements of [output, weight, input_indices]
+        # first we always have replicate all for inputs and output
+        placement_list: List[Placement] = [Replicate()] * 3
```

Comment:
Is there anything special about the placements of `[Replicate()] * 3`? If not shall we consider renaming it to something like `all_replicate` to be consistent with `colwise_sharding` and `batch_sharding`?

Reply:
Yeah sounds good, nothing special, I can rename them